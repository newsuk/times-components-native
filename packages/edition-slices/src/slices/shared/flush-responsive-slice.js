import { useResponsiveContext } from "@times-components-native/responsive";
import { editionBreakpoints } from "@times-components-native/styleguide";

const FlushResponsiveSlice = ({
  renderSmall,
  renderMedium,
  renderWide,
  renderHuge,
}) => {
  const { editionBreakpoint, orientation } = useResponsiveContext();
  switch (editionBreakpoint) {
    case editionBreakpoints.small:
      return renderSmall(editionBreakpoint, orientation);
    case editionBreakpoints.medium:
      return renderMedium(editionBreakpoint, orientation);
    case editionBreakpoints.wide:
      return (
        (renderWide && renderWide(editionBreakpoint, orientation)) ||
        (renderMedium && renderMedium(editionBreakpoint, orientation))
      );
    case editionBreakpoints.huge:
      return (
        (renderHuge && renderHuge(editionBreakpoint, orientation)) ||
        (renderWide && renderWide(editionBreakpoint, orientation)) ||
        (renderMedium && renderMedium(editionBreakpoint, orientation))
      );
    default:
      return renderSmall(editionBreakpoint, orientation);
  }
};

export default FlushResponsiveSlice;
