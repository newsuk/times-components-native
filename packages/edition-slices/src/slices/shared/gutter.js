import React from "react";
import PropTypes from "prop-types";
import { View } from "react-native";
import { useResponsiveContext } from "@times-components-native/responsive";
import styleFactory from "./styles";

const Gutter = ({ children, grow }) => {
  const { isTablet, editionBreakpoint } = useResponsiveContext();
  const styles = styleFactory(isTablet, editionBreakpoint);

  return (
    <View style={[styles.gutterStyles, grow && { flexGrow: 1 }]}>
      {children}
    </View>
  );
};

Gutter.propTypes = {
  children: PropTypes.node.isRequired,
};

export default Gutter;
