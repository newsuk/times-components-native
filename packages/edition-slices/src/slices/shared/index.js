import Gutter from "./gutter";
import ContentWrapper from "./content-wrapper";
import ResponsiveSlice from "./responsive-slice";
import FlushResponsiveSlice from "./flush-responsive-slice";

export { Gutter, ContentWrapper, ResponsiveSlice, FlushResponsiveSlice };
