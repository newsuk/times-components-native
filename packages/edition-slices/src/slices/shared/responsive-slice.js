import React from "react";
import PropTypes from "prop-types";
import { useResponsiveContext } from "@times-components-native/responsive";
import { editionBreakpoints } from "@times-components-native/styleguide";
import Gutter from "./gutter";
import ContentWrapper from "./content-wrapper";

const ResponsiveSlice = ({
  renderSmall,
  renderMedium,
  renderWide,
  renderHuge,
  grow = false,
}) => {
  const { editionBreakpoint, orientation } = useResponsiveContext();
  switch (editionBreakpoint) {
    case editionBreakpoints.medium:
      return (
        <Gutter grow={grow}>
          {renderMedium(editionBreakpoint, orientation)}
        </Gutter>
      );
    case editionBreakpoints.wide:
      return (
        <Gutter grow={grow}>
          {(renderWide && renderWide(editionBreakpoint, orientation)) ||
            (renderMedium && renderMedium(editionBreakpoint, orientation))}
        </Gutter>
      );
    case editionBreakpoints.huge:
      return (
        <Gutter grow={grow}>
          <ContentWrapper>
            {(renderHuge && renderHuge(editionBreakpoint, orientation)) ||
              (renderWide && renderWide(editionBreakpoint, orientation)) ||
              (renderMedium && renderMedium(editionBreakpoint, orientation))}
          </ContentWrapper>
        </Gutter>
      );
    case editionBreakpoints.small:
    default:
      return <>{renderSmall(editionBreakpoint, orientation)}</>;
  }
};

ResponsiveSlice.propTypes = {
  renderSmall: PropTypes.func.isRequired,
  renderMedium: PropTypes.func.isRequired,
  renderWide: PropTypes.func,
  renderHuge: PropTypes.func,
};

ResponsiveSlice.defaultProps = {
  renderWide: null,
  renderHuge: null,
};

export default ResponsiveSlice;
