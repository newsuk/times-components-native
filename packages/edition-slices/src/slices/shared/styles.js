import {
  editionMaxWidth,
  sliceContentMaxWidth,
  spacing,
  editionBreakpoints,
} from "@times-components-native/styleguide";

const tabletPaddingHorizontalResolver = {
  [editionBreakpoints.small]: 0,
  // [editionBreakpoints.smallTablet]: spacing(2),
  [editionBreakpoints.smallTablet]: 0,
  [editionBreakpoints.medium]: spacing(2),
  [editionBreakpoints.wide]: spacing(6),
  [editionBreakpoints.huge]: spacing(2),
};

export default (isTablet, breakpoint) => ({
  contentWrapperStyles: {
    flex: 1,
    alignSelf: "center",
    width: sliceContentMaxWidth,
  },

  gutterStyles: {
    alignSelf: "center",
    maxWidth: "100%",
    paddingHorizontal: isTablet
      ? tabletPaddingHorizontalResolver[breakpoint]
      : 0,
    width: editionMaxWidth,
  },
});
