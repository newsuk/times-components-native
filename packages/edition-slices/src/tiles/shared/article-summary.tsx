import React, { useEffect, useState, ReactNode } from "react";
import { Animated, StyleProp, TextStyle, ViewStyle, View } from "react-native";
import {
  default as ArticleSummaryComponent,
  ArticleSummaryContent,
  ArticleSummaryHeadline,
  ArticleSummaryStrapline,
} from "@times-components-native/article-summary";
import {
  BylineInput,
  Markup,
  Tile,
} from "@times-components-native/fixture-generator/src/types";

import { ArticleFlags } from "@times-components-native/article-flag";
import {
  colours,
  ARTICLE_READ_ANIMATION,
} from "@times-components-native/styleguide/index";
import PositionedTileStar from "./positioned-tile-star";

type ArticleRead = {
  id: string;
  highlight: boolean;
};

type ArticleReadState = {
  read: boolean;
  animate: boolean;
};

type MarkAsReadProps = {
  articleReadState: ArticleReadState;
  children: ReactNode;
  opacityAnimation: Animated.Value;
  opacity: number;
};

interface Props {
  bylines?: BylineInput[];
  bylineStyle?: StyleProp<ViewStyle>;
  bylineOnTop: boolean;
  flagColour?: any;
  flagsStyle?: StyleProp<ViewStyle>;
  headlineStyle?: StyleProp<TextStyle>;
  labelColour?: string;
  linesOfTeaserToRender?: number;
  strapline?: string;
  straplineStyle?: StyleProp<ViewStyle>;
  style?: StyleProp<ViewStyle>;
  summary?: Markup;
  summaryStyle?: any;
  tile: Tile;
  withStar?: boolean;
  underneathTextStar?: boolean;
  centeredStar?: boolean;
  isDarkStar?: boolean;
  isTablet: boolean;
  readArticles: [ArticleRead] | null;
  starStyle?: StyleProp<ViewStyle>;
  hideLabel?: boolean;
  whiteSpaceHeight?: number;
}

export const MarkAsRead = ({
  children,
  articleReadState,
  opacityAnimation,
  opacity,
}: MarkAsReadProps) => (
  <>
    {articleReadState.animate ? (
      <Animated.View
        style={{
          opacity: opacityAnimation,
        }}
      >
        {children}
      </Animated.View>
    ) : articleReadState.read ? (
      <View style={{ opacity }}>{children}</View>
    ) : (
      children
    )}
  </>
);

const ArticleSummary: React.FC<Props> = ({
  bylines,
  bylineStyle,
  bylineOnTop = false,
  flagColour = {},
  flagsStyle,
  headlineStyle,
  labelColour,
  linesOfTeaserToRender,
  strapline,
  straplineStyle,
  style,
  summary,
  summaryStyle,
  readArticles,
  tile,
  withStar = true,
  whiteSpaceHeight,
  underneathTextStar = false,
  centeredStar = false,
  isDarkStar = false,
  isTablet = false,
  starStyle,
  hideLabel = false,
}) => {
  const {
    headline: tileHeadline,
    article: {
      expirableFlags,
      longRead,
      hasVideo,
      headline,
      shortHeadline,
      label,
      section,
      id,
    },
  } = tile;

  const [standardOpacity] = useState(new Animated.Value(1));
  const [straplineOpacity] = useState(new Animated.Value(1));
  const [summaryOpacity] = useState(new Animated.Value(1));

  const sharedTimingConfig = {
    delay: ARTICLE_READ_ANIMATION.DELAY,
    duration: ARTICLE_READ_ANIMATION.DURATION,
    useNativeDriver: true,
  };

  const articleReadOpacity = {
    standard: 0.57,
    summary: 0.7,
  };

  const getArticleReadState = (
    isTablet: boolean,
    readArticles: Array<ArticleRead> | null,
    articleId: string,
  ): ArticleReadState => ({
    read: isTablet && !!readArticles?.some((obj) => obj.id === articleId),
    animate:
      isTablet &&
      !!readArticles?.some((obj) => obj.highlight && obj.id === articleId),
  });

  const articleReadState = getArticleReadState(isTablet, readArticles, id);

  useEffect(() => {
    Animated.parallel([
      Animated.timing(standardOpacity, {
        ...sharedTimingConfig,
        toValue: articleReadOpacity.standard,
      }),
      Animated.timing(straplineOpacity, {
        ...sharedTimingConfig,
        toValue: articleReadOpacity.standard,
      }),
      Animated.timing(summaryOpacity, {
        ...sharedTimingConfig,
        toValue: articleReadOpacity.summary,
      }),
    ]).start();
  }, [articleReadState]);

  const renderContent = (articleReadState: ArticleReadState) => (
    <MarkAsRead
      articleReadState={articleReadState}
      opacityAnimation={summaryOpacity}
      opacity={articleReadOpacity.summary}
    >
      <ArticleSummaryContent
        ast={summary}
        style={summaryStyle}
        lineHeight={(summaryStyle && summaryStyle.lineHeight) || undefined}
        whiteSpaceHeight={whiteSpaceHeight}
        initialLines={linesOfTeaserToRender}
      />
    </MarkAsRead>
  );

  const renderFlags = (articleReadState: ArticleReadState) => (
    <MarkAsRead
      articleReadState={articleReadState}
      opacityAnimation={standardOpacity}
      opacity={articleReadOpacity.standard}
    >
      <ArticleFlags
        {...flagColour}
        style={flagsStyle}
        flags={expirableFlags}
        longRead={longRead}
      />
    </MarkAsRead>
  );

  const renderSaveStar = () => (
    <PositionedTileStar
      articleId={id}
      isDarkStar={isDarkStar}
      centeredStar={centeredStar}
      underneathTextStar={underneathTextStar}
      style={starStyle}
    />
  );

  const renderHeadline = (articleReadState: ArticleReadState) => (
    <MarkAsRead
      articleReadState={articleReadState}
      opacityAnimation={standardOpacity}
      opacity={articleReadOpacity.standard}
    >
      <ArticleSummaryHeadline
        headline={tileHeadline || shortHeadline || headline || ""}
        style={headlineStyle}
      />
    </MarkAsRead>
  );

  const renderStrapline = (articleReadState: ArticleReadState) =>
    strapline && (
      <MarkAsRead
        articleReadState={articleReadState}
        opacityAnimation={straplineOpacity}
        opacity={articleReadOpacity.standard}
      >
        <ArticleSummaryStrapline strapline={strapline} style={straplineStyle} />
      </MarkAsRead>
    );

  return (
    <ArticleSummaryComponent
      articleReadState={articleReadState}
      bylineProps={bylines ? { ast: bylines, bylineStyle, bylineOnTop } : null}
      content={summary && renderContent(articleReadState)}
      flags={renderFlags(articleReadState)}
      headline={renderHeadline(articleReadState)}
      labelProps={{
        color:
          labelColour ||
          (section && colours.section[section]) ||
          colours.section.default,
        isVideo: hasVideo,
        title: label,
        hide: hideLabel,
      }}
      strapline={renderStrapline(articleReadState)}
      saveStar={withStar && renderSaveStar()}
      style={style}
    />
  );
};

export default ArticleSummary;
