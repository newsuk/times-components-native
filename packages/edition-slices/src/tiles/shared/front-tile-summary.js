import React, { Component } from "react";
import PropTypes from "prop-types";
import {
  ArticleSummaryHeadline,
  ArticleSummaryStrapline,
} from "@times-components-native/article-summary";

import {
  FrontArticleSummary,
  FrontArticleSummaryContent,
} from "@times-components-native/front-page";

class FrontTileSummary extends Component {
  constructor(props) {
    super(props);
    this.renderContent = this.renderContent.bind(this);
    this.renderHeadline = this.renderHeadline.bind(this);
    this.renderStrapline = this.renderStrapline.bind(this);
  }

  renderContent() {
    const {
      summary,
      summaryStyle,
      whiteSpaceHeight,
      linesOfTeaserToRender,
    } = this.props;

    return (
      <FrontArticleSummaryContent
        ast={summary}
        style={summaryStyle}
        whiteSpaceHeight={whiteSpaceHeight}
        initialLines={linesOfTeaserToRender}
      />
    );
  }

  renderHeadline() {
    const {
      tile: {
        headline: tileHeadline,
        article: { headline, shortHeadline },
      },
      headlineStyle,
    } = this.props;

    return (
      <ArticleSummaryHeadline
        headline={tileHeadline || shortHeadline || headline}
        style={headlineStyle}
      />
    );
  }

  renderStrapline() {
    const { strapline, straplineStyle } = this.props;

    return (
      <ArticleSummaryStrapline strapline={strapline} style={straplineStyle} />
    );
  }

  render() {
    const { bylines, bylineStyle, strapline, style, summary } = this.props;

    return (
      <FrontArticleSummary
        bylineProps={bylines ? { ast: bylines, bylineStyle } : null}
        content={summary ? this.renderContent() : undefined}
        headline={this.renderHeadline()}
        strapline={strapline ? this.renderStrapline() : undefined}
        style={style}
      />
    );
  }
}

FrontTileSummary.propTypes = {
  bylineStyle: PropTypes.shape({}),
  headlineStyle: PropTypes.shape({}),
  strapline: PropTypes.string,
  straplineStyle: PropTypes.shape({}),
  style: PropTypes.shape({}),
  summary: PropTypes.arrayOf(PropTypes.shape({})),
  summaryStyle: PropTypes.shape({}),
  tile: PropTypes.shape({}).isRequired,
  bylines: PropTypes.shape({}).isRequired,
};

FrontTileSummary.defaultProps = {
  bylineStyle: null,
  headlineStyle: null,
  strapline: null,
  straplineStyle: null,
  style: null,
  summary: null,
  summaryStyle: null,
};

export default FrontTileSummary;
