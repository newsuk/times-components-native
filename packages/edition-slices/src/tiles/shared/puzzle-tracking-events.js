import { withTrackEvents } from "@times-components-native/tracking";

export default (Component) =>
  withTrackEvents(Component, {
    analyticsEvents: [
      {
        actionName: "Pressed",
        eventName: "onPress",
        getAttrs: ({ id, title, url }) => ({
          puzzleId: id,
          puzzleTitle: title,
          puzzleUrl: url,
          tileName: "Puzzle",
        }),
        trackingName: "Tile",
      },
    ],
  });
