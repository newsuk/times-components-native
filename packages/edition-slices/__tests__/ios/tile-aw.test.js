import shared from "../tile-aw/shared-tile-aw";

shared();
