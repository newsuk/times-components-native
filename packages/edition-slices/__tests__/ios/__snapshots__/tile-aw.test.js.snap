// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tile aw landscape huge 1`] = `
<Link
  linkStyle={
    Object {
      "flex": 1,
      "paddingBottom": 10,
      "paddingRight": 10,
    }
  }
>
  <Image
    aspectRatio={1.7777777777777777}
    style={
      Object {
        "width": "100%",
      }
    }
    uri="https://www.thetimes.co.uk/imageserver/image/methode%2Ftimes%2Fprod%2Fweb%2Fbin%2Fd3abdbfc-1776-11e6-b4ba-d249b128bacc.jpg?crop=620%2C348%2C0%2C32"
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F0F0F0",
        "flex": 1,
        "justifyContent": "center",
        "paddingBottom": 10,
        "paddingHorizontal": 60,
        "paddingTop": 30,
      }
    }
  >
    <View
      style={
        Object {
          "marginBottom": 5,
        }
      }
    >
      <ArticleLabel
        color="#005B8D"
        hide={false}
        isVideo={false}
        markAsRead={false}
        title="label"
      />
    </View>
    <Text
      accessibilityRole="header"
      aria-level="3"
      style={
        Object {
          "color": "#1D1D1B",
          "fontFamily": "TimesModern-Regular",
          "fontSize": 45,
          "fontWeight": "normal",
          "includeFontPadding": false,
          "lineHeight": 45,
          "marginBottom": 10,
          "textAlign": "center",
        }
      }
    >
      This is tile headline
    </Text>
    <ArticleFlags
      flags={
        Array [
          Object {
            "expiryTime": "2030-03-14T12:00:00.000Z",
            "type": "EXCLUSIVE",
          },
        ]
      }
      longRead={true}
    />
  </View>
</Link>
`;

exports[`tile aw landscape medium 1`] = `
<Link
  linkStyle={
    Object {
      "flex": 1,
      "paddingBottom": 10,
      "paddingRight": 10,
    }
  }
>
  <Image
    aspectRatio={1.7777777777777777}
    style={
      Object {
        "width": "100%",
      }
    }
    uri="https://www.thetimes.co.uk/imageserver/image/methode%2Ftimes%2Fprod%2Fweb%2Fbin%2Fd3abdbfc-1776-11e6-b4ba-d249b128bacc.jpg?crop=620%2C348%2C0%2C32"
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F0F0F0",
        "flex": 1,
        "justifyContent": "center",
        "paddingBottom": 10,
        "paddingHorizontal": 60,
        "paddingTop": 30,
      }
    }
  >
    <View
      style={
        Object {
          "marginBottom": 5,
        }
      }
    >
      <ArticleLabel
        color="#005B8D"
        hide={false}
        isVideo={false}
        markAsRead={false}
        title="label"
      />
    </View>
    <Text
      accessibilityRole="header"
      aria-level="3"
      style={
        Object {
          "color": "#1D1D1B",
          "fontFamily": "TimesModern-Regular",
          "fontSize": 30,
          "fontWeight": "normal",
          "includeFontPadding": false,
          "lineHeight": 30,
          "marginBottom": 10,
          "textAlign": "center",
        }
      }
    >
      This is tile headline
    </Text>
    <ArticleFlags
      flags={
        Array [
          Object {
            "expiryTime": "2030-03-14T12:00:00.000Z",
            "type": "EXCLUSIVE",
          },
        ]
      }
      longRead={true}
    />
  </View>
</Link>
`;

exports[`tile aw landscape wide 1`] = `
<Link
  linkStyle={
    Object {
      "flex": 1,
      "paddingBottom": 10,
      "paddingRight": 10,
    }
  }
>
  <Image
    aspectRatio={1.7777777777777777}
    style={
      Object {
        "width": "100%",
      }
    }
    uri="https://www.thetimes.co.uk/imageserver/image/methode%2Ftimes%2Fprod%2Fweb%2Fbin%2Fd3abdbfc-1776-11e6-b4ba-d249b128bacc.jpg?crop=620%2C348%2C0%2C32"
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F0F0F0",
        "flex": 1,
        "justifyContent": "center",
        "paddingBottom": 10,
        "paddingHorizontal": 60,
        "paddingTop": 30,
      }
    }
  >
    <View
      style={
        Object {
          "marginBottom": 5,
        }
      }
    >
      <ArticleLabel
        color="#005B8D"
        hide={false}
        isVideo={false}
        markAsRead={false}
        title="label"
      />
    </View>
    <Text
      accessibilityRole="header"
      aria-level="3"
      style={
        Object {
          "color": "#1D1D1B",
          "fontFamily": "TimesModern-Regular",
          "fontSize": 35,
          "fontWeight": "normal",
          "includeFontPadding": false,
          "lineHeight": 35,
          "marginBottom": 10,
          "textAlign": "center",
        }
      }
    >
      This is tile headline
    </Text>
    <ArticleFlags
      flags={
        Array [
          Object {
            "expiryTime": "2030-03-14T12:00:00.000Z",
            "type": "EXCLUSIVE",
          },
        ]
      }
      longRead={true}
    />
  </View>
</Link>
`;
